File: python/mirage/nfs/fs.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import errno
import os
import posixpath
import stat as stat_bits

from mirage.mount.core import MountCore
from mirage.nfs.config import NFSConfig
from mirage.nfs.errors import StaleHandleError
from mirage.nfs.ids import ROOT_PATH, IdTable
from mirage.nfs.types import DirEntry, NFSAttrs, SetAttrs
from mirage.nfs.writebuf import WriteBuffer
from mirage.ops import Ops

# The core slices to the end when asked for more than the file holds,
# which is how a whole-file read is spelled through a sized API.
_WHOLE_FILE = 1 << 62


def _join(parent: str, name: str) -> str:
    return posixpath.join(parent,
                          name) if parent != ROOT_PATH else ROOT_PATH + name


class MirageNFS:
    """The NFSv3 filesystem the server crate calls back into.

    One method per trait callback, each one async so it runs on the
    workspace event loop and reaches the op door the same way a shell
    command does: mount grants, admission policies, cache and namespace
    all fire once, at that door. The adapter itself owns only what the
    protocol needs and mirage does not have -- which file id names which
    path, and the writes a client has sent but not yet had stored.

    Paths crossing this boundary are mount-relative; the mount prefix is
    applied by the op facade this is constructed with.
    """

    def __init__(self, ops: Ops, config: NFSConfig | None = None) -> None:
        self._ops = ops
        # The shared mount core. Every filesystem semantic the two
        # kernel tiers agree on -- link display, macOS metadata, entry
        # naming, stat shaping, the size-unknown rules -- is decided
        # there, so the adapters cannot drift. What stays here is what
        # NFSv3 alone needs: ids, the write buffer, the wire attrs.
        self._core = MountCore(ops)
        self._config = config or NFSConfig()
        self._ids = IdTable()
        self._writes = WriteBuffer()
        # One lock per file that has been written; dropped with the
        # buffer it guards, so the table tracks live files rather than
        # every id ever minted.
        self._flush_locks: dict[int, asyncio.Lock] = {}
        self._root = self._ids.alloc(ROOT_PATH)

    def root_dir(self) -> int:
        """The file id of the export root."""
        return self._root

    async def lookup(self, dirid: int, name: str) -> int:
        """Resolve a name inside a directory to a file id.

        Args:
            dirid (int): the parent directory's id.
            name (str): the entry name.

        Returns:
            int: the entry's file id.

        Raises:
            StaleHandleError: the parent id is unknown.
            FileNotFoundError: no such entry.
        """
        path = _join(self._ids.resolve(dirid), name)
        # getattr refuses macOS metadata names and reports a link as a
        # link rather than following it, both of which this repeated.
        await self._core.getattr(path)
        return self._ids.alloc(path)

    async def getattr(self, fileid: int) -> NFSAttrs:
        """Attributes for a file id, counting writes not yet stored.

        Args:
            fileid (int): the file to stat.

        Returns:
            NFSAttrs: the shape the Rust layer converts to fattr3.
        """
        return await self._entry_attrs(fileid, self._ids.resolve(fileid))

    async def read(self, fileid: int, offset: int, count: int) -> bytes:
        """Read through any writes still buffered for this file.

        Args:
            fileid (int): the file to read.
            offset (int): where the read starts.
            count (int): how many bytes the client asked for.

        Returns:
            bytes: the slice, short at end of file.
        """
        path = self._ids.resolve(fileid)
        base = await self._read_base(path)
        return self._writes.overlay(fileid, base, offset, count)

    async def write(self, fileid: int, offset: int, data: bytes) -> NFSAttrs:
        """Buffer a write and answer with the size the client expects.

        The bytes are stored on flush, not here: this server answers
        every write as durable and never forwards a COMMIT, so the
        adapter batches and bounds the window itself.

        Args:
            fileid (int): the file being written.
            offset (int): byte offset the client wrote at.
            data (bytes): the payload.

        Returns:
            NFSAttrs: post-write attributes, with the extended size.
        """
        path = self._ids.resolve(fileid)
        full = self._writes.append(fileid,
                                   offset,
                                   data,
                                   max_bytes=self._config.max_buffered_bytes)
        if full:
            await self._flush_one(fileid, path)
        return await self._entry_attrs(fileid, path)

    async def create(self, dirid: int, name: str) -> int:
        """Create an empty file and return its id.

        Args:
            dirid (int): the parent directory's id.
            name (str): the new file's name.

        Returns:
            int: the new file's id.
        """
        path = _join(self._ids.resolve(dirid), name)
        fh = await self._core.create(path)
        # NFSv3 is handle-free; the core's handle would leak.
        await self._core.release(fh)
        return self._ids.alloc(path)

    async def mkdir(self, dirid: int, name: str) -> int:
        """Create a directory and return its id.

        Args:
            dirid (int): the parent directory's id.
            name (str): the new directory's name.

        Returns:
            int: the new directory's id.
        """
        path = _join(self._ids.resolve(dirid), name)
        await self._core.mkdir(path)
        return self._ids.alloc(path)

    async def remove(self, dirid: int, name: str) -> None:
        """Remove a file or directory.

        The server routes both REMOVE and RMDIR here, so the entry is
        stat-ed first to pick the right op. Buffered writes are dropped
        rather than flushed: storing them would bring the file back.

        Args:
            dirid (int): the parent directory's id.
            name (str): the entry to remove.
        """
        path = _join(self._ids.resolve(dirid), name)
        fileid = self._ids.id_for(path)
        if fileid is not None:
            self._writes.drop(fileid)
            self._flush_locks.pop(fileid, None)
        # The core's unlink unlinks a link rather than following it, and
        # its getattr reports a link as a link, which keeps one out of
        # the rmdir arm and lets a broken one be removed at all.
        attrs = await self._core.attrs_for(path)
        if stat_bits.S_ISDIR(attrs["st_mode"]):
            await self._core.rmdir(path)
        else:
            await self._core.unlink(path)
        if fileid is not None:
            self._ids.invalidate(fileid)

    async def rename(self, from_dirid: int, from_name: str, to_dirid: int,
                     to_name: str) -> None:
        """Move an entry, carrying its id and pending writes with it.

        Pending writes are flushed to the old path first: they were
        acknowledged against it, and flushing after the move would merge
        them onto whatever now lives at the destination.

        Args:
            from_dirid (int): source directory id.
            from_name (str): source entry name.
            to_dirid (int): destination directory id.
            to_name (str): destination entry name.

        Raises:
            OSError: EINVAL when the destination lies inside the
                source, refused before the backend is touched.
        """
        src = _join(self._ids.resolve(from_dirid), from_name)
        dst = _join(self._ids.resolve(to_dirid), to_name)
        self._ids.guard_rename(src, dst)
        fileid = self._ids.id_for(src)
        if fileid is not None and self._writes.has_pending(fileid):
            await self._flush_one(fileid, src)
        await self._core.rename(src, dst)
        self._ids.rename(src, dst)

    async def setattr(self, fileid: int, attrs: SetAttrs) -> NFSAttrs:
        """Apply the one settable attribute: size.

        mode, uid, gid and the timestamps are accepted and discarded,
        exactly as the FUSE adapter does -- a mirage backend has nowhere
        to persist them, and refusing would fail ordinary tools.

        Args:
            fileid (int): the file to change.
            attrs (SetAttrs): requested change; only ``size`` acts.

        Returns:
            NFSAttrs: attributes after the change.
        """
        path = self._ids.resolve(fileid)
        size = attrs.size
        if size is not None:
            self._writes.clip(fileid, size)
            await self._core.truncate(path, size)
        return await self._entry_attrs(fileid, path)

    async def set_size(self, fileid: int, size: int | None) -> NFSAttrs:
        """The wire layer's SETATTR entry point, on primitives.

        The Rust boundary crosses on primitives, so it calls this
        rather than constructing a :class:`SetAttrs`.

        Args:
            fileid (int): the file to change.
            size (int | None): new length, or None when the request
                carried no size and everything is discarded.

        Returns:
            NFSAttrs: attributes after the change.
        """
        return await self.setattr(fileid, SetAttrs(size=size))

    async def symlink(self, dirid: int, name: str, target: str) -> int:
        """Create a symlink and return its id.

        Args:
            dirid (int): the parent directory's id.
            name (str): the link's name.
            target (str): what the link points at, stored verbatim.

        Returns:
            int: the link's file id.
        """
        path = _join(self._ids.resolve(dirid), name)
        await self._core.symlink(path, target)
        return self._ids.alloc(path)

    async def readlink(self, fileid: int) -> str:
        """The target a symlink holds.

        Args:
            fileid (int): the link's file id.

        Returns:
            str: the target as the client should see it -- relative
            targets verbatim, absolute ones rewritten relative to the
            link's directory, since an absolute target names a virtual
            path the client would otherwise resolve against its own
            root and escape the mount.

        Raises:
            OSError: EINVAL when the id does not name a link.
        """
        path = self._ids.resolve(fileid)
        target = self._core.link_target(path)
        if target is None:
            raise OSError(errno.EINVAL, os.strerror(errno.EINVAL), path)
        return target

    async def readdir(self,
                      dirid: int,
                      cookie: int = 0,
                      max_entries: int | None = None) -> list[DirEntry]:
        """List a directory, resuming after the entry ``cookie`` names.

        The cookie is the last-seen entry's fileid: the server crate
        derives the wire cookie from each entry's id and hands it back
        as ``start_after``. Resume keys on identity, never on comparing
        magnitudes -- ids are minted in access order, so a later entry
        may carry a smaller id than an earlier one.

        Args:
            dirid (int): the directory to list.
            cookie (int): fileid of the last entry seen; 0 starts at
                the top.
            max_entries (int | None): cap on entries returned.

        Returns:
            list[DirEntry]: name, fileid, cookie and attributes, with
            ``cookie == fileid`` on every entry.
        """
        path = self._ids.resolve(dirid)
        # The facade answers in paths -- a child mount with a trailing
        # slash -- so names are derived the way MountCore.readdir does,
        # and macOS metadata names are dropped the same way.
        # The core derives names from the facade's paths and drops
        # macOS metadata; "." and ".." are libfuse's to emit, and NFSv3
        # carries them in the reply header instead.
        names = [
            n for n in await self._core.readdir(path) if n not in (".", "..")
        ]
        entries: list[DirEntry] = []
        resuming = cookie != 0
        for name in names:
            child = _join(path, name)
            fileid = self._ids.alloc(child)
            if resuming:
                if fileid == cookie:
                    resuming = False
                continue
            entries.append(
                DirEntry(name=name,
                         fileid=fileid,
                         cookie=fileid,
                         attrs=await self._entry_attrs(fileid, child)))
            if max_entries is not None and len(entries) >= max_entries:
                break
        return entries

    async def flush(self, fileid: int) -> None:
        """Store one file's buffered writes.

        Args:
            fileid (int): the file to flush.
        """
        if self._writes.has_pending(fileid):
            await self._flush_one(fileid, self._ids.resolve(fileid))

    async def flush_all(self) -> None:
        """Store every buffered write.

        Used by the idle sweep and at teardown. A file id that went
        stale under a pending buffer is dropped rather than raised: one
        dead entry must not stop the rest from being stored.
        """
        for fileid in self._writes.pending_ids():
            await self._flush_or_drop(fileid)

    async def flush_idle(self) -> None:
        """Store writes untouched for longer than the idle window."""
        for fileid in self._writes.idle_ids(self._config.idle_flush_seconds):
            await self._flush_or_drop(fileid)

    async def _flush_or_drop(self, fileid: int) -> None:
        """Flush one file id, discarding its writes if the id went stale.

        A sweep covers every buffered id, so one entry whose path is
        gone must not stop the others from being stored.

        Args:
            fileid (int): the file to flush.
        """
        try:
            path = self._ids.resolve(fileid)
        except StaleHandleError:
            self._writes.drop(fileid)
            self._flush_locks.pop(fileid, None)
            return
        await self._flush_one(fileid, path)

    def _flush_lock(self, fileid: int) -> asyncio.Lock:
        """The lock serializing one file's flushes.

        Kept here rather than on ``WriteBuffer``: the state holders are
        await-free by design, and a lock inside one would not help
        anyway -- what has to be atomic spans a read, a take and a
        write, which only the caller can bracket.

        Args:
            fileid (int): the file whose flushes to serialize.

        Returns:
            asyncio.Lock: the lock for that file.
        """
        lock = self._flush_locks.get(fileid)
        if lock is None:
            lock = asyncio.Lock()
            self._flush_locks[fileid] = lock
        return lock

    async def _flush_one(self, fileid: int, path: str) -> None:
        """Store one file's buffered writes, one flush at a time.

        Read, take and write are one critical section. Without it two
        flushes of the same file -- an idle timer against a size
        trigger, or either against teardown -- each read the same stored
        base and take different batches, and whichever store lands last
        drops the other batch. The client was told those bytes were
        durable.

        Args:
            fileid (int): the file to flush.
            path (str): its path, resolved by the caller.
        """
        async with self._flush_lock(fileid):
            base = await self._read_base(path)
            pending = self._writes.take(fileid)
            if not pending:
                return
            await self._ops.write(path, WriteBuffer.merge(base, pending))

    async def _read_base(self, path: str) -> bytes:
        try:
            return await self._core.read(path, _WHOLE_FILE, 0, None)
        except (FileNotFoundError, IsADirectoryError):
            return b""

    async def _entry_attrs(self, fileid: int, path: str) -> NFSAttrs:
        """The wire attributes for one entry, over the core's POSIX ones.

        The core decides what the entry *is* -- a link reported as a
        link rather than followed, a size-unknown file reported as 0,
        macOS metadata refused -- and this converts that answer into the
        three facts NFSv3 puts on the wire, plus the size a client
        should see, which counts writes buffered but not yet stored.

        ``attrs_for`` rather than ``getattr``: the id was minted by a
        lookup that already applied the name policy, and re-applying it
        here would disappear an entry the client just created.

        Args:
            fileid (int): the entry's id.
            path (str): the entry's mount-relative path.
        """
        entry = await self._core.attrs_for(path)
        mode = entry["st_mode"]
        is_dir = bool(stat_bits.S_ISDIR(mode))
        size = 0 if is_dir else self._writes.pending_size(
            fileid, entry["st_size"])
        return NFSAttrs(fileid=fileid,
                        size=size,
                        is_dir=is_dir,
                        is_symlink=bool(stat_bits.S_ISLNK(mode)),
                        mode=mode & 0o7777,
                        mtime_epoch=float(entry["st_mtime"]))

File: typescript/packages/node/src/nfs/fs.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { posix } from 'node:path'

import type { Ops } from '@struktoai/mirage-core/ops/ops'
import { FileType } from '@struktoai/mirage-core/types'
import { mtimeMs } from '@struktoai/mirage-core/utils/stat_view'
import type { FileStat } from '@struktoai/mirage-core/types'
import { sortedByCodePoints } from '@struktoai/mirage-core/utils/sort'

import { errnoError } from '../mount/errors.ts'
import { isMacosMetadata } from '../mount/platform/macos.ts'
import { NFSConfig } from './config.ts'
import { StaleHandleError } from './errors.ts'
import { IdTable, ROOT_PATH } from './ids.ts'
import type { DirEntry, NFSAttrs } from './types.ts'
import { WriteBuffer } from './writebuf.ts'

function joinPath(parent: string, name: string): string {
  return parent === ROOT_PATH ? ROOT_PATH + name : posix.join(parent, name)
}

function enoent(path: string): Error {
  return errnoError('ENOENT', `no such file or directory: ${path}`)
}

function einval(path: string): Error {
  return errnoError('EINVAL', `invalid argument: ${path}`)
}

/**
 * The NFSv3 filesystem the server crate calls back into.
 *
 * One method per trait callback, each one async so it runs on the
 * event loop and reaches the op door the same way a shell command
 * does: mount grants, admission policies, cache and namespace all
 * fire once, at that door. The adapter itself owns only what the
 * protocol needs and mirage does not have -- which file id names
 * which path, and the writes a client has sent but not yet had
 * stored.
 *
 * Paths crossing this boundary are mount-relative; the mount prefix
 * is applied by the op facade this is constructed with.
 */
export class MirageNFS {
  private readonly ops: Ops
  private readonly config: NFSConfig
  private readonly ids = new IdTable()
  private readonly writes = new WriteBuffer()
  // One chain per file that has been written, cleared with the buffer it
  // guards, so the map tracks live files rather than every id minted.
  private readonly flushChains = new Map<number, Promise<void>>()
  private readonly root: number

  constructor(ops: Ops, config: NFSConfig = new NFSConfig()) {
    this.ops = ops
    this.config = config
    this.root = this.ids.alloc(ROOT_PATH)
  }

  /** The file id of the export root. */
  rootDir(): number {
    return this.root
  }

  /** Resolve a name inside a directory to a file id. */
  async lookup(dirid: number, name: string): Promise<number> {
    if (isMacosMetadata(name)) {
      // Finder and Spotlight probe these on every listing; answering
      // here keeps the probe off the backend, as MountCore does.
      throw enoent(name)
    }
    const path = joinPath(this.ids.resolve(dirid), name)
    if (this.linkTarget(path) === null) await this.ops.stat(path)
    return this.ids.alloc(path)
  }

  /** Attributes for a file id, counting writes not yet stored. */
  async getattr(fileid: number): Promise<NFSAttrs> {
    return this.entryAttrs(fileid, this.ids.resolve(fileid))
  }

  /** Read through any writes still buffered for this file. */
  async read(fileid: number, offset: number, count: number): Promise<Buffer> {
    const path = this.ids.resolve(fileid)
    const base = await this.readBase(path)
    return this.writes.overlay(fileid, base, offset, count)
  }

  /**
   * Buffer a write and answer with the size the client expects.
   *
   * The bytes are stored on flush, not here: this server answers
   * every write as durable and never forwards a COMMIT, so the
   * adapter batches and bounds the window itself.
   */
  async write(fileid: number, offset: number, data: Buffer): Promise<NFSAttrs> {
    const path = this.ids.resolve(fileid)
    const full = this.writes.append(fileid, offset, data, this.config.maxBufferedBytes)
    if (full) await this.flushOne(fileid, path)
    return this.entryAttrs(fileid, path)
  }

  /** Create an empty file and return its id. */
  async create(dirid: number, name: string): Promise<number> {
    const path = joinPath(this.ids.resolve(dirid), name)
    await this.ops.create(path)
    return this.ids.alloc(path)
  }

  /** Create a directory and return its id. */
  async mkdir(dirid: number, name: string): Promise<number> {
    const path = joinPath(this.ids.resolve(dirid), name)
    await this.ops.mkdir(path)
    return this.ids.alloc(path)
  }

  /**
   * Remove a file or directory.
   *
   * The server routes both REMOVE and RMDIR here, so the entry is
   * stat-ed first to pick the right op. A link is unlinked whatever
   * it points at: stat would follow it, and following a link to a
   * directory would rmdir the target instead of the link. Buffered
   * writes are dropped rather than flushed -- storing them would
   * bring the file back.
   */
  async remove(dirid: number, name: string): Promise<void> {
    const path = joinPath(this.ids.resolve(dirid), name)
    const fileid = this.ids.idFor(path)
    if (this.linkTarget(path) !== null) {
      if (fileid !== undefined) this.dropBuffered(fileid)
      await this.ops.unlink(path)
      if (fileid !== undefined) this.ids.invalidate(fileid)
      return
    }
    const stat = await this.ops.stat(path)
    if (fileid !== undefined) this.dropBuffered(fileid)
    if (stat.type === FileType.DIRECTORY) {
      await this.ops.rmdir(path)
    } else {
      await this.ops.unlink(path)
    }
    if (fileid !== undefined) this.ids.invalidate(fileid)
  }

  /**
   * Move an entry, carrying its id and pending writes with it.
   *
   * Pending writes are flushed to the old path first: they were
   * acknowledged against it, and flushing after the move would merge
   * them onto whatever now lives at the destination.
   */
  async rename(
    fromDirid: number,
    fromName: string,
    toDirid: number,
    toName: string,
  ): Promise<void> {
    const src = joinPath(this.ids.resolve(fromDirid), fromName)
    const dst = joinPath(this.ids.resolve(toDirid), toName)
    this.ids.guardRename(src, dst)
    const fileid = this.ids.idFor(src)
    if (fileid !== undefined && this.writes.hasPending(fileid)) {
      await this.flushOne(fileid, src)
    }
    await this.ops.rename(src, dst)
    this.ids.rename(src, dst)
  }

  /**
   * Apply the one settable attribute: size.
   *
   * mode, uid, gid and the timestamps are accepted and discarded,
   * exactly as the FUSE adapter does -- a mirage backend has nowhere
   * to persist them, and refusing would fail ordinary tools.
   */
  async setSize(fileid: number, size: number | null): Promise<NFSAttrs> {
    const path = this.ids.resolve(fileid)
    if (size !== null) {
      this.writes.clip(fileid, size)
      await this.ops.truncate(path, size)
    }
    return this.entryAttrs(fileid, path)
  }

  /** Create a symlink and return its id. */
  async symlink(dirid: number, name: string, target: string): Promise<number> {
    const path = joinPath(this.ids.resolve(dirid), name)
    await this.ops.symlink(path, target)
    return this.ids.alloc(path)
  }

  /**
   * The target a symlink holds, as the client should see it.
   *
   * Relative targets are returned verbatim; absolute ones name
   * virtual paths and are rewritten relative to the link's directory,
   * since a client would otherwise resolve them against its own root
   * and escape the mount.
   */
  // A link is namespace state, so this reads no backend; the callback
  // contract is still one async method per NFS procedure.
  // eslint-disable-next-line @typescript-eslint/require-await
  async readlink(fileid: number): Promise<string> {
    const path = this.ids.resolve(fileid)
    const target = this.linkTarget(path)
    if (target === null) throw einval(path)
    return target
  }

  /**
   * List a directory, resuming after the entry `cookie` names.
   *
   * The cookie is the last-seen entry's fileid: the server crate
   * derives the wire cookie from each entry's id and hands it back as
   * `startAfter`. Resume keys on identity, never on comparing
   * magnitudes -- ids are minted in access order, so a later entry
   * may carry a smaller id than an earlier one.
   */
  async readdir(dirid: number, cookie = 0, maxEntries?: number): Promise<DirEntry[]> {
    const path = this.ids.resolve(dirid)
    // The facade answers in paths -- a child mount with a trailing
    // slash -- so names are derived the way MountCore.readdir does,
    // and macOS metadata names are dropped the same way.
    const found = new Set<string>()
    for (const entry of await this.ops.readdir(path)) {
      const part = entry.replace(/\/+$/, '').split('/').pop() ?? ''
      if (part !== '' && !isMacosMetadata(part)) found.add(part)
    }
    const entries: DirEntry[] = []
    let resuming = cookie !== 0
    for (const name of sortedByCodePoints(found)) {
      const child = joinPath(path, name)
      const fileid = this.ids.alloc(child)
      if (resuming) {
        if (fileid === cookie) resuming = false
        continue
      }
      entries.push({
        name,
        fileid,
        cookie: fileid,
        attrs: await this.entryAttrs(fileid, child),
      })
      if (maxEntries !== undefined && entries.length >= maxEntries) break
    }
    return entries
  }

  /** Store one file's buffered writes. */
  async flush(fileid: number): Promise<void> {
    if (this.writes.hasPending(fileid)) {
      await this.flushOne(fileid, this.ids.resolve(fileid))
    }
  }

  /**
   * Store every buffered write. Used at teardown. A file id that went
   * stale under a pending buffer is dropped rather than raised: one
   * dead entry must not stop the rest from being stored.
   */
  async flushAll(): Promise<void> {
    for (const fileid of this.writes.pendingIds()) await this.flushOrDrop(fileid)
  }

  /** Store writes untouched for longer than the idle window. */
  async flushIdle(): Promise<void> {
    for (const fileid of this.writes.idleIds(this.config.idleFlushSeconds)) {
      await this.flushOrDrop(fileid)
    }
  }

  private async flushOrDrop(fileid: number): Promise<void> {
    let path: string
    try {
      path = this.ids.resolve(fileid)
    } catch (err) {
      if (!(err instanceof StaleHandleError)) throw err
      this.writes.drop(fileid)
      this.flushChains.delete(fileid)
      return
    }
    await this.flushOne(fileid, path)
  }

  /**
   * Store one file's buffered writes, one flush at a time.
   *
   * Read, take and store are one critical section. Without it two
   * flushes of the same file -- an idle timer against a size trigger,
   * or either against teardown -- each read the same stored base and
   * take different batches, and whichever store settles last drops the
   * other batch. The client was told those bytes were durable.
   *
   * Serialized by chaining rather than a lock, which JavaScript has no
   * need of: the chain is the queue, and a failed flush does not strand
   * the ones behind it.
   */
  /** Forget a file's pending writes and the chain that serialized them. */
  private dropBuffered(fileid: number): void {
    this.writes.drop(fileid)
    this.flushChains.delete(fileid)
  }

  private async flushOne(fileid: number, path: string): Promise<void> {
    const previous = this.flushChains.get(fileid) ?? Promise.resolve()
    const mine = previous.catch(() => undefined).then(() => this.storeOne(fileid, path))
    this.flushChains.set(fileid, mine)
    try {
      await mine
    } finally {
      // Only the tail clears itself: an earlier link finishing must not
      // drop a chain a later flush is still queued behind.
      if (this.flushChains.get(fileid) === mine) this.flushChains.delete(fileid)
    }
  }

  private async storeOne(fileid: number, path: string): Promise<void> {
    const base = await this.readBase(path)
    const pending = this.writes.take(fileid)
    if (pending.length === 0) return
    await this.ops.writeFile(path, WriteBuffer.merge(base, pending))
  }

  private async readBase(path: string): Promise<Buffer> {
    try {
      return Buffer.from(await this.ops.readFile(path, { raw: true }))
    } catch {
      // missing file or a directory: the write starts from empty
      return Buffer.alloc(0)
    }
  }

  /**
   * The target to present for a namespace link, null otherwise.
   *
   * The link check must precede any ops stat, exactly as in
   * MountCore.getattr: the op facade follows namespace links, so a
   * stat on a link path reports the target and the link itself
   * becomes invisible.
   */
  private linkTarget(path: string): string | null {
    const links = this.ops.links
    if (links === null) return null
    const target = links.readlink(path)
    if (target === null) return null
    if (!target.startsWith('/')) return target
    const parent = path.split('/').slice(0, -1).join('/') || '/'
    return posix.relative(parent, target)
  }

  /** Attributes for one entry, seeing a link as itself. */
  private async entryAttrs(fileid: number, path: string): Promise<NFSAttrs> {
    const target = this.linkTarget(path)
    if (target !== null) {
      return {
        fileid,
        size: Buffer.byteLength(target),
        isDir: false,
        isSymlink: true,
      }
    }
    return this.attrs(fileid, await this.ops.stat(path))
  }

  private attrs(fileid: number, stat: FileStat): NFSAttrs {
    const isDir = stat.type === FileType.DIRECTORY
    const size = isDir ? 0 : this.writes.pendingSize(fileid, stat.size ?? 0)
    const out: NFSAttrs = {
      fileid,
      size,
      isDir,
      isSymlink: stat.type === FileType.SYMLINK,
    }
    if (stat.mode !== null) out.mode = stat.mode
    // The wire needs an nfstime3 and vfs.rs reads exactly this field; a
    // row with no time leaves it unset, and the client reads 1970, which
    // is honest where a fabricated "now" would not be.
    const ms = mtimeMs(stat)
    if (ms !== null) out.mtimeEpoch = ms / 1000
    return out
  }
}
